torch/_functorch/aot_autograd.py

```diff
@@ -138,86 +138,365 @@ def posthook(grad_input, grad_output):
 class InputAliasInfo:
-    # This object gives us information about how to perform a metadata-mutation
-    # on original_fwd_inputs[base_idx]
-    #   (This is an index into the inputs *before* we make synthetic bases)
-    base_idx: int
-    # sizes, strides and storage offset of the aliased output are all returned as actual (sym)ints
-    # in the compiled forward. These indices tell us where in the forward outputs to grab them.
-    sizes_idx: int
-    strides_idx: int
-    storage_offset_idx: int
-    # We store the actual output alias that we traced in the forward (should be a fake tensor)
-    # to grab any other non-symbolic properties on the output alias, like requires_grad.
-    tensor_meta: Tensor
+    mutation_type: MutationType
 
 # This class encapsulates all aliasing + mutation info we need about the forward graph
 # See a more detailed overview of the edge case handling at
 # https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit
 @dataclass(frozen=True)
 class ViewAndMutationMeta:
-    # length: # user forward inputs
-    # For every input, tells us whether the input:
-    # (a) is not mutated
-    # (b) only metadata is mutated
-    # (c) data (and maybe metadta) is mutated
-    mutated_input_info: List[MutationType]
-    # length: (# inputs of the user forward)
-    # metadata_mutation_input_info[i] is not None <====> mutated_input_info[i] == MutationType.metadata_only
-    # We stash the updated FakeTensor that we traced with in the forward in here,
-    # that way we can use it to replay the metadata mutation
-    metadata_mutation_input_info: List[Optional[InputAliasInfo]]
-    # length: # outputs in the compiled forward (not including output alias symints). Equal to:
-    # length: (# inputs w data mutations) + (# outputs that don't alias inputs)
+    # length = # user inputs
+    # This gives us info about every input, and what sort of mutation happened to it (if any)
+    input_info: List[InputAliasInfo]
+    # length = # user outputs + # intermediates that are returned as graph outputs
+    # This gives us info about every output (mostly around whether it aliases other tensors)
+    output_info: List[OutputAliasInfo]
+    # length = # outputs in the compiled forward. Equal to:
+    #   # mutated inps + # user outputs + # intermediate bases
+    #   (same as)
+    #   len(x for x in input_info if x.MutationType != MutationType.none) + len(output_info) + num_intermediate_bases
     # For every output *and* mutated input returned from the forward,
     # tells us whether or not the output should require gradients or not
-    requires_grad_out_info: List[bool]
-    # length: # fw outputs
-    aliased_output_info: List[OutputAliasInfo]
-
-def gen_alias_from_base(aliased_base_tensor, size, stride, storage_offset, target_meta_tensor):
-    # handle R2C and C2R
+    requires_grad_info: List[bool]
+    # length = len(x for x in input_info if x.MutationType != MutationType.none)
```

Comment:
Instead of giving the length, why don't you just outright say how you computed it, it's basically this but in a few more lines.

Even better: why don't you just directly compute these lists in the `__post_init__` of this class, then there is no reason to give a big algorithm description, it's right below.